pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/flux/git"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	lister "github.com/fluxcd/helm-operator/pkg/client/listers/helm.fluxcd.io/v1"
+)
+
+// Various (final) errors.
+var (
+	ErrReleasesForMirror = errors.New("failed to get HelmRelease resources for mirror")
+	ErrNoMirror          = errors.New("no existing git mirror found")
+	ErrMirrorSync        = errors.New("failed syncing git mirror")
+)
+
+// ReleaseQueue is an add-only `workqueue.RateLimitingInterface`
+type ReleaseQueue interface {
+	AddRateLimited(item interface{})
+}
+
+// GitConfig holds the configuration for git operations.
+type GitConfig struct {
+	GitTimeout      time.Duration
+	GitPollInterval time.Duration
+}
+
+// GitChartDir holds a `git.Export` of a chart repository,
+// and the commit hash of the HEAD of this export.
+type GitChartDir struct {
+	*git.Export
+	Head string
+}
+
+func (d GitChartDir) ChartPath(relativePath string) string {
+	if d.Export != nil {
+		return filepath.Join(d.Dir(), relativePath)
+	}
+	return ""
+}
+
+// GitChartSync syncs `sourceRef`s with their mirrors, and queues
+// updates for `v1.HelmRelease`s the sync changes are relevant for.
+type GitChartSync struct {
+	logger log.Logger
+	config GitConfig
+
+	lister lister.HelmReleaseLister
+
+	mirrors *git.Mirrors
+
+	sourcesMu sync.RWMutex
+	sources   map[string]sourceRef
+
+	releaseQueue ReleaseQueue
+}
+
+// sourceRef is used for book keeping, so that we know when a
+// signal we receive from a mirror is an actual update for a
+// release, and if the source we hold is still the one referred
+// to in the `v1.HelmRelease`.
+type sourceRef struct {
+	mirror string
+	remote string
+	ref    string
+	head   string
+}
+
+// forHelmRelease returns true if the given `v1.HelmRelease`s
+// `v1.GitChartSource` matches the sourceRef.
+func (c sourceRef) forHelmRelease(hr *v1.HelmRelease) bool {
+	if hr == nil || hr.Spec.GitChartSource == nil {
+		return false
+	}
+	return c.mirror == mirrorName(hr) && c.remote == hr.Spec.GitURL && c.ref == hr.Spec.Ref
+}
+
+func NewGitChartSync(logger log.Logger,
+	lister lister.HelmReleaseLister, cfg GitConfig, queue ReleaseQueue) *GitChartSync {
+
+	return &GitChartSync{
+		logger:       logger,
+		config:       cfg,
+
+		lister:       lister,
+
+		mirrors:      git.NewMirrors(),
+		sources:      make(map[string]sourceRef),
+
+		releaseQueue: queue,
+	}
+}
+
+// Run starts the mirroring of git repositories, and processes mirror
+// changes on signal, scheduling a release for a `HelmRelease` resource
+// when the update is relevant to the release.
+func (c *GitChartSync) Run(stopCh <-chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
```

Comment:
Why not place this code directly in the `case <-stopCh:` instead of using a deferred function call?

Reply:
One of the few original lines from the original code.

Historically looking back, I think this is because the previous `panic` possibility, which would still make the `defer` happen (but would not reach the `stopCh`).